### PR TITLE
[blpapi] add new port

### PR DESCRIPTION
**Pull request by sweemer**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

Disclaimer: I am not affiliated with Bloomberg in any way. I am simply adding this port based on their publicly available libraries.

I have two questions:

1. Their license looks like MIT but I don't think I should be the one to say for sure. Can we leave the license declaration blank until we can get somebody from Bloomberg to clarify?
2. Bloomberg only provide dynamic libraries, so I have reflected that in the `supports` field; however, I would like to remove that restriction and [force dynamic linkage ](https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_library_linkage) even when the user is using a static triplet. Can you guide me on how to do that?

### CODE REVIEW

File: ports/blpapi/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
+    }
+  ]
+}
```
Comment (JonLiu1993) [inline ports/blpapi/vcpkg.json:17]: @sweemer, Thanks for your PR,` vcpkg_cmake_configure` and `vcpkg_cmake_install` provided by `vcpkg-cmake` port
`vcpkg_cmake_config_fixup` provided by `vcpkg-cmake-config` port, blpapi does not use these functions, so "vcpkg-cmake" and "vcpkg-cmake-config" should not be used

File: ports/blpapi/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
+    }
+  ]
```
Comment (JonLiu1993) [inline ports/blpapi/vcpkg.json:16]:
```suggestion
```

Comment (JonLiu1993) [reply ports/blpapi/vcpkg.json:16]: @sweemer, Thanks for your PR, vcpkg_cmake_configure and vcpkg_cmake_install provided by vcpkg-cmake port
vcpkg_cmake_config_fixup provided by vcpkg-cmake-config port, blpapi does not use these functions, so "vcpkg-cmake" and "vcpkg-cmake-config" should not be used

**Comment by JonLiu1993** [comment]
Tested usage successfully by `blpapi:x64-windows`:
````
The package blpapi provides CMake targets:

    find_package(blpapi CONFIG REQUIRED)
    target_link_libraries(main PRIVATE blpapi)
````

**Comment by JonLiu1993** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

File: ports/blpapi/vcpkg.json (hunk 2 above)
Comment (sweemer) [reply ports/blpapi/vcpkg.json]: Done

**Review by JonLiu1993 — DISMISSED** [review]

File: ports/blpapi/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+elseif ("${VCPKG_TARGET_ARCHITECTURE}" MATCHES "x64")
+    set(BITS_SUFFIX 64)
+else()
+    message(FATAL_ERROR "Unrecognized architecture.")
+endif()
+
+if (VCPKG_TARGET_IS_LINUX)
+    file(GLOB SO_FILES LIST_DIRECTORIES false "${SOURCE_PATH}/Linux/*${BITS_SUFFIX}.so")
+else()
+    file(GLOB DLL_FILES LIST_DIRECTORIES false "${SOURCE_PATH}/lib/*${BITS_SUFFIX}.dll")
```
Comment (BillyONeal) [inline ports/blpapi/portfile.cmake:35]: I checked that the DLLs don't appear to link with a CRT so this debug release mismatch might be OK.

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
>Their license looks like MIT but I don't think I should be the one to say for sure. Can we leave the license declaration blank until we can get somebody from Bloomberg to clarify?

LGTM

>Bloomberg only provide dynamic libraries, so I have reflected that in the supports field; however, I would like to remove that restriction and [force dynamic linkage](https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_library_linkage) even when the user is using a static triplet. Can you guide me on how to do that?

It would literally just be removing that block from the supports clause. The problem is that dynamic linkage with static CRT is usually doom so we intentionally don't make that easy to do out of the box.

**Comment by BillyONeal** [comment]
Please double check that you're happy with nitpicks I fixed, and if so, indicate that you are happy by marking this 'Ready for review'

**Comment by sweemer** [comment]
@BillyONeal I'm happy with the changes you've made.

Regarding forced dynamic linkage, do I understand correctly that you don't recommend forcing it with a static triplet? If so, then I will leave it the way it is. But if it's acceptable in this case then I would like to remove that part of the supports clause before merging. If the concern is about dynamic linkage with static CRT then I don't mind changing it to something like `(linux | (windows & !uwp)) & !(!static & staticcrt) & (x86 | x64)`.

**Comment by BillyONeal** [comment]
> Regarding forced dynamic linkage, do I understand correctly that you don't recommend forcing it with a static triplet?

You can force dynamic linkage with a static triplet *if* the resulting DLL doesn't speak about the standard library at all. Most libraries aren't prepared to deal with, for example, `std::vector&`s that you can't use because they belong to a totally different CRT.

There are global variables like the debug lock or the locale lock which cause doom if you try to touch standard library components visibly across such an interface. When you dynamically link with the CRT, those global variables live in the CRT's DLLs, but a static triplet requests a static CRT. Static CRT plus dynamic linkage means each DLL has its own CRT, and thus its own copy of those globals, and thus can't contain standard library types in their interface at all.

**Comment by apollomok** [comment]
Is the path out-updated?  It cannot download from the path.

error: https://bcms.bloomberg.com/BLPAPI-Generic/blpapi_cpp_3.20.2.2-windows.zip: WinHttpSendRequest failed with exit code 12007

**Comment by sweemer** [comment]
Yes, it might need to be updated. I can take a look next week.

**Comment by sweemer** [comment]
Updated in https://github.com/microsoft/vcpkg/pull/40130

**Pull request merged**